File: android-app/app 2/src/main/java/com/bandite/sonicwalkscape/data/models/Tour.kt
package com.bandite.sonicwalkscape.data.models

import com.google.android.gms.maps.model.LatLng
import com.google.gson.annotations.SerializedName

data class Tour(
    val id: String,
    val slug: String,
    val title: Map<String, String>,
    val descriptionPreview: Map<String, String>,
    val completionMessage: Map<String, String>? = null,
    val busInfo: Map<String, String>? = null,
    val city: String,
    val durationMinutes: Int,
    val distanceKm: Double,
    @SerializedName("difficulty")
    val difficultyRaw: String = "facile",
    val languages: List<String>,
    val isProtected: Boolean,
    @SerializedName("imageUrl")
    val coverImageUrl: String? = null,
    val coverTrailerUrl: String? = null,
    val routePolyline: String? = null,
    var points: List<TourPoint> = emptyList(),
    var isDownloaded: Boolean = false
) {
    // Display properties with language fallback
    fun getDisplayTitle(language: String = "en"): String {
        return title[language] ?: title.values.firstOrNull() ?: "Untitled Tour"
    }

    fun getDisplayDescription(language: String = "en"): String {
        return descriptionPreview[language] ?: descriptionPreview.values.firstOrNull() ?: ""
    }

    fun getDisplayCompletionMessage(language: String = "en"): String? {
        return completionMessage?.get(language) ?: completionMessage?.values?.firstOrNull()
    }

    fun getDisplayBusInfo(language: String = "en"): String? {
        return busInfo?.get(language) ?: busInfo?.values?.firstOrNull()
    }

    // Duration in seconds
    val durationSeconds: Long
        get() = durationMinutes * 60L

    // Distance in meters
    val distanceMeters: Double
        get() = distanceKm * 1000

    val difficulty: Difficulty
        get() = when (difficultyRaw) {
            "medio" -> Difficulty.MODERATE
            "difficile" -> Difficulty.CHALLENGING
            else -> Difficulty.EASY
        }

    // Full image URL with base URL prepended
    fun getFullCoverImageUrl(baseUrl: String): String? {
        val url = coverImageUrl ?: return null
        return if (url.startsWith("http://") || url.startsWith("https://")) {
            url
        } else {
            "$baseUrl$url"
        }
    }

    // Full trailer URL with base URL prepended
    fun getFullCoverTrailerUrl(baseUrl: String): String? {
        val url = coverTrailerUrl ?: return null
        return if (url.startsWith("http://") || url.startsWith("https://")) {
            url
        } else {
            "$baseUrl$url"
        }
    }

    // Parse route polyline string into coordinates
    // Format: "lat1,lng1;lat2,lng2;lat3,lng3"
    val routeCoordinates: List<LatLng>
        get() {
            val polyline = routePolyline ?: return emptyList()
            if (polyline.isEmpty()) return emptyList()

            return polyline.split(";").mapNotNull { pair ->
                val components = pair.split(",")
                if (components.size == 2) {
                    val lat = components[0].toDoubleOrNull()
                    val lng = components[1].toDoubleOrNull()
                    if (lat != null && lng != null) {
                        LatLng(lat, lng)
                    } else null
                } else null
            }
        }

    enum class Difficulty {
        EASY,
        MODERATE,
        CHALLENGING
    }

    enum class Category {
        HISTORY,
        NATURE,
        ART,
        ARCHITECTURE,
        FOOD,
        CULTURE
    }
}

File: mobile-app/ios/SonicWalkscape/SonicWalkscape/Models/Tour.swift
import Foundation
import CoreLocation

struct Tour: Codable, Identifiable, Equatable {
    let id: String
    let slug: String
    let title: [String: String]
    let descriptionPreview: [String: String]
    let completionMessage: [String: String]?
    let busInfo: [String: String]?
    let city: String
    let durationMinutes: Int
    let distanceKm: Double
    let difficultyRaw: String
    let languages: [String]
    let isProtected: Bool
    let coverImageUrl: String?
    let coverTrailerUrl: String?
    let routePolyline: String?
    var points: [TourPoint] = []
    var isDownloaded: Bool = false

    // Computed properties for backward compatibility
    var displayTitle: String {
        title["en"] ?? title.values.first ?? "Untitled Tour"
    }

    var displayDescription: String {
        descriptionPreview["en"] ?? descriptionPreview.values.first ?? ""
    }

    var displayCompletionMessage: String? {
        completionMessage?["en"] ?? completionMessage?.values.first
    }

    var displayBusInfo: String? {
        busInfo?["en"] ?? busInfo?.values.first
    }

    var duration: TimeInterval {
        TimeInterval(durationMinutes * 60)
    }

    var distance: Double {
        distanceKm * 1000
    }

    // Keep category for UI filtering
    var category: Category {
        .history // Default, can be enhanced later
    }

    var difficulty: Difficulty {
        switch difficultyRaw {
        case "medio": return .moderate
        case "difficile": return .challenging
        default: return .easy
        }
    }

    var displayDifficulty: String {
        LocalizedStrings.shared.localizedDifficulty(difficultyRaw)
    }

    // Full image URL with base URL prepended
    var fullCoverImageUrl: String? {
        guard let coverImageUrl = coverImageUrl else { return nil }
        // If already absolute URL, return as is
        if coverImageUrl.hasPrefix("http://") || coverImageUrl.hasPrefix("https://") {
            return coverImageUrl
        }
        // Otherwise prepend base URL
        return "\(Constants.API.baseURL)\(coverImageUrl)"
    }

    // Full trailer URL with base URL prepended
    var fullCoverTrailerUrl: String? {
        guard let coverTrailerUrl = coverTrailerUrl else { return nil }
        // If already absolute URL, return as is
        if coverTrailerUrl.hasPrefix("http://") || coverTrailerUrl.hasPrefix("https://") {
            return coverTrailerUrl
        }
        // Otherwise prepend base URL
        return "\(Constants.API.baseURL)\(coverTrailerUrl)"
    }

    // Parse route polyline string into coordinates
    // Format: "lat1,lng1;lat2,lng2;lat3,lng3"
    var routeCoordinates: [CLLocationCoordinate2D] {
        guard let routePolyline = routePolyline, !routePolyline.isEmpty else {
            return []
        }

        return routePolyline.split(separator: ";").compactMap { pair in
            let components = pair.split(separator: ",")
            guard components.count == 2,
                  let lat = Double(components[0]),
                  let lng = Double(components[1]) else {
                return nil
            }
            return CLLocationCoordinate2D(latitude: lat, longitude: lng)
        }
    }

    enum Difficulty: String, Codable {
        case easy
        case moderate
        case challenging
    }

    enum Category: String, Codable {
        case history
        case nature
        case art
        case architecture
        case food
        case culture
    }

    enum CodingKeys: String, CodingKey {
        case id, slug, title, descriptionPreview, completionMessage, busInfo, city, durationMinutes, distanceKm, languages, isProtected
        case difficultyRaw = "difficulty"
        case coverImageUrl = "imageUrl"
        case coverTrailerUrl
        case routePolyline
    }

    // Manual initializer for creating Tour objects in code
    init(
        id: String,
        slug: String,
        title: [String: String],
        descriptionPreview: [String: String],
        completionMessage: [String: String]? = nil,
        busInfo: [String: String]? = nil,
        city: String,
        durationMinutes: Int,
        distanceKm: Double,
        difficultyRaw: String = "facile",
        languages: [String],
        isProtected: Bool,
        coverImageUrl: String? = nil,
        coverTrailerUrl: String? = nil,
        routePolyline: String? = nil,
        points: [TourPoint] = [],
        isDownloaded: Bool = false
    ) {
        self.id = id
        self.slug = slug
        self.title = title
        self.descriptionPreview = descriptionPreview
        self.completionMessage = completionMessage
        self.busInfo = busInfo
        self.city = city
        self.durationMinutes = durationMinutes
        self.distanceKm = distanceKm
        self.difficultyRaw = difficultyRaw
        self.languages = languages
        self.isProtected = isProtected
        self.coverImageUrl = coverImageUrl
        self.coverTrailerUrl = coverTrailerUrl
        self.routePolyline = routePolyline
        self.points = points
        self.isDownloaded = isDownloaded
    }

    // Custom decoder
    init(from decoder: Decoder) throws {
        let container = try decoder.container(keyedBy: CodingKeys.self)

        id = try container.decode(String.self, forKey: .id)
        slug = try container.decode(String.self, forKey: .slug)
        title = try container.decode([String: String].self, forKey: .title)
        descriptionPreview = try container.decode([String: String].self, forKey: .descriptionPreview)
        completionMessage = try container.decodeIfPresent([String: String].self, forKey: .completionMessage)
        busInfo = try container.decodeIfPresent([String: String].self, forKey: .busInfo)
        city = try container.decode(String.self, forKey: .city)
        durationMinutes = try container.decode(Int.self, forKey: .durationMinutes)
        distanceKm = try container.decode(Double.self, forKey: .distanceKm)
        difficultyRaw = try container.decodeIfPresent(String.self, forKey: .difficultyRaw) ?? "facile"
        languages = try container.decode([String].self, forKey: .languages)
        isProtected = try container.decode(Bool.self, forKey: .isProtected)
        coverImageUrl = try container.decodeIfPresent(String.self, forKey: .coverImageUrl)
        coverTrailerUrl = try container.decodeIfPresent(String.self, forKey: .coverTrailerUrl)
        routePolyline = try container.decodeIfPresent(String.self, forKey: .routePolyline)

        points = []
        isDownloaded = false
    }

    // Custom encoder to match the decoder
    func encode(to encoder: Encoder) throws {
        var container = encoder.container(keyedBy: CodingKeys.self)

        try container.encode(id, forKey: .id)
        try container.encode(slug, forKey: .slug)
        try container.encode(title, forKey: .title)
        try container.encode(descriptionPreview, forKey: .descriptionPreview)
        try container.encodeIfPresent(completionMessage, forKey: .completionMessage)
        try container.encodeIfPresent(busInfo, forKey: .busInfo)
        try container.encode(city, forKey: .city)
        try container.encode(durationMinutes, forKey: .durationMinutes)
        try container.encode(distanceKm, forKey: .distanceKm)
        try container.encode(difficultyRaw, forKey: .difficultyRaw)
        try container.encode(languages, forKey: .languages)
        try container.encode(isProtected, forKey: .isProtected)
        try container.encodeIfPresent(coverImageUrl, forKey: .coverImageUrl)
        try container.encodeIfPresent(coverTrailerUrl, forKey: .coverTrailerUrl)
        try container.encodeIfPresent(routePolyline, forKey: .routePolyline)
    }

    // Manual Equatable implementation (required because we have custom Codable)
    static func == (lhs: Tour, rhs: Tour) -> Bool {
        return lhs.id == rhs.id &&
               lhs.slug == rhs.slug &&
               lhs.city == rhs.city &&
               lhs.durationMinutes == rhs.durationMinutes &&
               lhs.distanceKm == rhs.distanceKm &&
               lhs.difficultyRaw == rhs.difficultyRaw &&
               lhs.isProtected == rhs.isProtected
    }
}
